### PR TITLE
idle: Add parser for cpu_idle events

### PR DESCRIPTION
**Pull request by bjackman**

### CODE REVIEW

**Comment by JaviMerino** [comment]
Please add the `trace.dat` instead of the `trace.txt`.  The existing `trace.txt`s make it hard to implement #210 (I had the code but I had to change the existing `trace.txt`s as `trace.dat` and I never got time to do it).  Let's not increase the number of them by having them in binary form.  You will have to annotate `TestCpuIdle` with:

``` python
@unittest.skipUnless(utils_tests.trace_cmd_installed(),
                     "trace-cmd not installed")
```

**Comment by JaviMerino** [comment]
travis says that this breaks the tests because `tests/trace_empty.txt` is no longer "empty".   Can you remove the `cpu_idle` events in there to avoid breaking the testsuite?

File: tests/test_idle.py [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+
+import os
+import sys
+
+import utils_tests
+import trappy
+
+sys.path.append(os.path.join(utils_tests.TESTS_DIRECTORY, "..", "trappy"))
+
+class BaseTestIdle(utils_tests.SetupDirectory):
```
Comment (JaviMerino) [inline tests/test_idle.py]: No need to create `BaseTestIdle` to be inherited by only one test.  Just move the `__init__` to `TestCpuIdle`.

File: tests/test_idle.py [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+class TestCpuIdle(BaseTestIdle):
+    def test_get_dataframe(self):
+        df = trappy.FTrace().cpu_idle.data_frame
+
+        exp_states = [2, -1, -1, 1, -1, 0, -1, 2, -1, 0, 0, -1, 0, -1, -1, -1, 2]
+        exp_cpus =   [5,  2,  1, 2,  5, 5,  3, 3,  0, 1, 0,  0, 0 , 0 , 5,  1, 5]
+
+        self.assertEqual(len(df), len(exp_states))
+
+        self.assertTrue((df["state"] == exp_states).all())
```
Comment (JaviMerino) [inline tests/test_idle.py]: If/When this fails, it gives an unhelpful message.  It's better to let unittest do the comparison, something like (untested):

``` python
self.assertEquals(df["state"].tolist(), exp_states)
```

Same for the `cpu_id` column.

File: tests/test_idle.py (hunk 1 above)
Comment (bjackman) [inline tests/test_idle.py]: Good point, will amend (I just copy pasted this from the sched tests, hence the strangeness)

File: trappy/idle.py [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+from trappy.dynamic import register_ftrace_parser
+
+class CpuIdle(Base):
+    """Parse cpu_idle"""
+
+    unique_word = "cpu_idle"
+
+    def finalize_object(self):
+        """The trace contains "4294967295" instead of "-1" when exiting an idle
+        state.
```
Comment (bjackman) [inline trappy/idle.py]: Changing this from a docstring to a comment

**Comment by bjackman** [comment]
Pushed a new version using `assert_series_equal`, unfortunately (?) you also have to explicitly do `assert_index_equal` too, `assert_series_equal` only seems to check the values.

File: tests/test_idle.py [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+@unittest.skipUnless(utils_tests.trace_cmd_installed(),
+                     "trace-cmd not installed")
+class TestCpuIdle(utils_tests.SetupDirectory):
+    def __init__(self, *args, **kwargs):
+        super(TestCpuIdle, self).__init__(
+            [("trace_idle.dat", "trace.dat")],
+            *args,
+            **kwargs)
+
+    def test_get_dataframe(self):
```
Comment (JaviMerino) [inline tests/test_idle.py]: Add a docstring for the function (just one line, like """cpu_idle events are parsed""").  `nosetests` will use it to describe the tests when running `nosetests -v` or when reporting failures.

**Comment by JaviMerino** [comment]
great, thanks! :heart:

**Pull request merged**